types/tx.go

```diff
@@ -158,3 +192,40 @@ func ComputeProtoSizeForTxs(txs []Tx) int64 {
 	pdData := data.ToProto()
 	return int64(pdData.Size())
 }
+
+// UnwrapMalleatedTx attempts to unmarshal the provided transaction into a malleated
+// transaction wrapper, if this an be done, then it returns true. A malleated
+// transaction is a normal transaction that has been derived (malleated) from a
+// different original transaction. The returned hash is that of the original
+// transaction, which allows us to remove the original transaction from the
+// mempool. NOTE: protobuf sometimes does not throw an error if the transaction
+// passed is not a tmproto.MalleatedTx, since the schema for PayForMessage is kept
+// in the app, we cannot perform further checks without creating an import
+// cycle.
+func UnwrapMalleatedTx(tx Tx) (malleatedTx tmproto.MalleatedTx, isMalleated bool) {
+	// attempt to unmarshal into a a malleated transaction
+	err := proto.Unmarshal(tx, &malleatedTx)
+	if err != nil {
+		return malleatedTx, false
+	}
+	// this check will fail to catch unwanted types should those unmarshalled
+	// types happen to have a hash sized slice of bytes in the same field number
+	// as originalTxHash. TODO(evan): either fix this, or better yet use a different
+	// mechanism
+	if len(malleatedTx.OriginalTxHash) != tmhash.Size {
```

Comment:
This is usually solved by embedding magic numbers (e.g., a field with a constant value, say uint64). That is a way more robust way to make sure we're not failing to detect an unwanted type.